xdsl/transforms/canonicalization_patterns/riscv.py

```diff
@@ -58,6 +58,31 @@ def match_and_rewrite(self, op: riscv.MulOp, rewriter: PatternRewriter) -> None:
             rewriter.replace_matched_op(riscv.MVOp(op.rs2, rd=rd))
 
 
+class MultiplyImmediatePowerOfTwo(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv.MulOp, rewriter: PatternRewriter) -> None:
+        if (log_rs1 := exact_log2(op.rs1)) is not None:
+            rd = cast(riscv.IntRegisterType, op.rd.type)
```

Comment:
It's a shame this cast is needed. It looks like a pyright bug to me after inspecting the types.